### DISCUSSION TITLE
The bot does not follow a direct order.

### DISCUSSION THREAD

**Discussion by pavel4win** [FakeFishGames/Barotrauma#14914] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

We were attacked by tigers.
They fell into a hatch, the hatch closed behind them and they began to suffocate without water. There were three of them.
And then a fearless drug addict doctor rushed at them with his sharpened screwdriver!
I ordered him to "follow me". He ignored my order, ran up to them and started poking with his screwdriver.
Of course they bit him and healed themselves.
What kind of nonsense is this?
Why does a bot attack creatures much stronger than it, with such a weak weapon?
Why did the bot ignore my order to "follow"
Why did the bot even rush at them? It did not have an active order to "defend".
It should have ignored distant enemies and avoided close enemies.
Why is there an order to "attack" among the possible orders for bots, but no order to "avoid danger"?

Addition:
I don't want the bot to heal me unless it has a direct order to "heal"
It doesn't spend meds efficiently.
Make the bot heal players only by direct order.
Let it heal  other bots without direct orders but NOT to heal player.
experienced player will take care of himself. Unless he directly orders the bot to "heal".

### Reproduction steps

[1.zip](https://github.com/user-attachments/files/17492138/1.zip)

Unzip the attached save.
Create a server, load a save and sail to the lighthouse.
Command the bot to "follow me"
The tigers will attack near the lighthouse or earlier.
Let them get inside.
And enjoy the suicidal stupidity of the bot.

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Multiplayer hosted from the in-game menu (= using a listen server)

### Version

v1.6.17.0 (Unto the Breach Update)

### Which operating system did you encounter this bug on?

Windows

**Comment by pavel4win** [comment]
Again faced with ignoring my orders.
The primary order was "to stand on the gun"
bot climbs to repair the superstructure where the hammerheads constantly hit.
It was not like that before
Looks as if the bots now consider what the AI ​​calculated for them to be a higher priority and not what the captain ordered.

And one more thing.
The diving suit is lying near the reactor, where there is no water or pressure. The bot climbs for the suit into the superstructure where there is pressure. It can't get there (the water flow throws it off) and it hangs.

Honestly, the game is turning into some kind of crazy farce. I'm not fighting Europe anymore. I'm fighting idiot bots.
The only consolation is that the idiot players behave much more stupidly than the bots. They are just complete idiots.

**Comment by pavel4win** [comment]
For example on the Orca sub, if there are more than four diving suits, then the bots, if they are ordered to clean up, for some reason drag all the extra diving suits into the fabricator room. Probably so that the first time the fabricator room is breached, the diving suits are carried out of the boat to the bottom of the ocean?
This is probably a special mockery of the players. The bots probably don't like that the players order them around?
Or is it a crooked AI?
Lets make it so that the bots don't drag diving suits if there are no free lockers for them. Let the diving suits lie where they were last taken off. That is, near the guns, near the reactor, near the boxes. In general, not in the fabricator room.
Please!
Game becomes annoying.

**Reply by SomeRandomNoobKekeke** [reply]
If you want bots not to touch your personal suit tell them to ignore it

> the diving suits are carried out of the boat to the bottom of the ocean?

https://github.com/user-attachments/assets/2993d09d-b553-4957-9871-f30ae25476aa

> for some reason drag all the extra diving suits into the fabricator room

https://github.com/user-attachments/assets/f2034289-0e45-46be-b537-87f9ead762a9

![john-cena-cena](https://github.com/user-attachments/assets/3b629856-756d-4157-9269-5c5c4dc53242)

**Reply by pavel4win** [reply]
Very funny. Did I write that the problem is in my personal suit? Of course I forbid touching my suit. They themselves are left without suits during a surprise attack and are forced to run to the fabricator for them. And if the fabricator is already breached and the suits floated away?
I don't need advice. I have long ago moved all lockers closer to the guns using the Barocraftables+ mod
I wrote this to draw devs attention to the lack AI ​​of the bots.
Or are you doing this on purpose so that life doesn't seem like honey to the players?
There are enough dangers on Europa without this. Bots idiots, that's are too much.
If I need idiots on a sub, I will launch an open server.

**Comment by pavel4win** [comment]
1. Maybe something has changed in game? In earlier versions of game I saw many times as suits used to fall out.
2. Suits lying on the bottom of the boat used to fall through the bottom if it was pierced. In your case, you just made an explosion. The Hammerhead hits from below, so everything that is lying there flies up into the air and then falls out of the boat through the bottom. Try this situation.

3. In version 1.5.9.2, the Bots carried all the extra COMBAT suits to the fabricator. Check with the combat suits. Maybe NOT after the "clean up" order, but in the absence of active orders.

4. You simply ignored the other glitches I pointed out in the behavior of the Bots? So, disobeying orders is normal?

**Reply by SomeRandomNoobKekeke** [reply]
I guess it was fixed in 1.4.3.0
![Screenshot_4](https://github.com/user-attachments/assets/1ed1e881-916e-4781-a790-b3d526d01e08)

https://github.com/user-attachments/assets/42e647ed-50ed-416d-be4b-4f6cef64188f

> the Bots carried all the extra COMBAT suits to the fabricator

I tried with combat suits, with and without cleaning order - same effect, they fill empty cabinets and then chill
And i've never seen such behaviour in vanilla in my life
Could Barocraftables+ mess up their cleaning tags?

> Bots with folow order fight monsters

How would you fix that? Do you want them to just stand and watch when you take them to caves or enemy outposts?
Or don't attack with melee? Or fight weak monsters and flee from strong? 

It's not that obvious what bots should do and there are much simpler fixes like :
giving all your bots some weapons,
don't let threshers in
Killing it yourself before poor bots get hurt

> bot climbs to repair the superstructure where the hammerheads constantly hit.

Did you mean supercapacitor?
They actually always did that
They fix capacitor themselves if it's broken and they can't shoot and no one wants to fix it
They also adjust recharge speed if they don't have enough energy

> I wrote this to draw devs attention to the lack AI ​​of the bots.
Or are you doing this on purpose so that life doesn't seem like honey to the players?

No, there's just 1.5 people fixing the code and they can't prioritize this
Also ai code is a mess and very hard to fix

**Comment by pavel4win** [comment]
"superstructure" - sorry for the misunderstanding.
This is the upper deck of the Orca, where the upper airlock and diving suits are. With glass roof so that the sailors wouldn't get bored when Hummerhead hits.
There is an upper airlock, it cannot be called a wheelhouse because there is no steering wheel. 

I am glad about the "suits fell out" fix. Thank you.

You convinced me. Perhaps the effect of suits accumulation in the fabricator was caused by the bot running into the fabricator, repairing the fabricator and throwing the suit right there.
My fabricator is turned into ballast, so the fabricator breaks constantly. During the round they dragged all the suits there.

About your question:
"How would you fix that? Do you want them to just stand and watch when you take them to caves or enemy outposts?
Or don't attack with melee? Or fight weak monsters and flee from strong?"

My IMHO is:

The bot must first of all follow the captain's orders. Even if it's suicide.
I believe that if a bot is ordered to "stand on the cannon" it must stand on the cannon. "Even if the ground is burning around" it. It should be distracted in only two cases.
1. The pressure increases - put on a suit, any other danger (fire, acid - avoid)
2. It is attacked: fight if he have a firearm or run away. Return as soon as the danger has passed.
Now Bots drop the cannons and start running around the ship waving screwdrivers as soon as one monster breaks inside. At the same time, they do not even shoot from their personal firearms.
I had several times that 6 bots with SMGs could not cope with 1 tiger. Because they did not shoot at it. Probably because of the fear of friendly fire.
I think bots should avoid danger if they do not have the order to "defend".
And they should definitely assess the strength and number of enemies and not rush into attack with a screwdriver against or 3 tigers.
If "estimating the strength and number of enemies" is too difficult, let them avoid all enemies until there is a firearm or a direct order to "defend".
And, repeat: captains order is primary!
In my case bot ignored even "follow" order!

I got the impression that bots have become dumber over the last six months.
Or I haven't played with bots on Orca.
For example, in earlier versions, if a bot was knocked down by a stream of water from a hatch, it would climb into the hatch again until it succeeded. My doctor (with permanent drug addict), when he climbs into the superstructure and is washed away by water, he stops and stands there doing nothing at all.

About your
"No, there's just 1.5 people fixing the code and they can't prioritize this
Also ai code is a mess and very hard to fix"

I was sure of it. After all, you are too kind to intentionally confuse the matter.
Maybe I should get hired and do AI bots?
I am a programmer and I have free time.

**Comment by itchyOwl** [comment]
Hey,

Thanks for your feedback.

Couldn't spot anything that would have changed/broken the behavior, when comparing the relevant changes between the summer update and the autumn update.

I agree that the bots should sometimes be more obedient, especially when they have the follow order. And I'm going to do some fixes to address this. Also seems that there are some issues/oversights regarding reactions to a friendly character getting attacked. However, I wasn't able to repro bots going full offensive when they shouldn't. It's intentional that they use whatever tools they have in the inventory for _defending_ themselves. I think I could make them try to flee more actively from the encounters, while in the defensive state though.

I don't think there should be a command for defending or surviving, as those behaviors are something the bots should be able to do autonomously. And it's intentional that surviving is prioritized over the orders, because otherwise bots would easily get killed while being ordered to do something.

As of the quality of the AI code, feel free to disagree or point out any issues, but in my opinon the human AI code is not at all messy, as it was refactored years ago, and we will continue to improve it. The code on monster AI is a different case, since it _wasn't_ refactored, and new features have been piled on a little bit shaky foundations. But in general it too is lot less buggy than it used to be.

It would help if you could give more detailed steps to repro some of the issues + maybe a video with debugai enabled.

**Reply by pavel4win** [reply]
In the save attached to this topic, the bot did not defend itself. It was away from the tigers, who were lying there suffocating. Tigers could not move without water and would have died. But bot rushed at them with a screwdriver.
I think bots should not attack strong creatures like tigers and raptors with a screwdriver. They should try to escape.

I am in no way criticizing your work. You guys are doing really cool stuff. I am very grateful to you for the opportunity to travel around Europe in a submarine.

I will try to collect more materials that will help you in improving the bots' AI.

**Reply by itchyOwl** [reply]
> In the save attached to this topic, the bot did not defend itself.

I opened the save, but it's unclear to me how to repro this scenario. I played for a couple of minutes and some monsters spawned, but they were not tigerthreshers (the spawns are random). So without further steps, it's difficult for me to repro what you encountered.

> I think bots should not attack strong creatures like tigers and raptors with a screwdriver. They should try to escape.

I agree and they shouldn't, but they may defend themselves and try to avoid the enemy. Sometimes the pathing to a safe place takes them close to the enemy (which hasn't been easy to fix), and when that happens, they may get stuck with a fight with the enemy. I could make them just flee instead of fighting, and they do that if they don't have any weapons. However, in many scenarios, it's better to fight with the enemy instead. Admittedly, in some, it's not.

The most recent changes should make them follow/hold position more obediently though. Be sure to test them in unstable, when they get merged (which should be soonish, unless QA finds issues in them).

> I will try to collect more materials that will help you in improving the bots' AI.

Thanks, appreciated!

**Comment by SomeRandomNoobKekeke** [comment]
@pavel4win https://discord.com/channels/103209684680323072/431606297134104577/1300495717336354847
![pat pat owl](https://github.com/user-attachments/assets/69292aed-07ab-4d7f-af80-86514048a887)

**Comment by itchyOwl** [comment]
Addressed at least some of the issues regarding bots not following the orders in FakeFishGames/Barotrauma-development#5971.

**Comment by pavel4win** [comment]
Regarding the fact that the bots carry diving suits to the engineering compartment on the Orca sub.
I am currently going through the campaign on the Orca and have become convinced of the following. The bots drag everything that is not nailed to the floor to the engineering compartment. That is, everything that cannot be put in the closet. These are diving suits of all types, empty and full boxes of shells for the coilgun. Nothing else I have leaving on the deck. This does not happen immediately, but gradually. At the end of any round, I find all these things in the engineering compartment under the stairs. I do not know where the bots got such a love for dragging everything there. But I have to order them to ignore all the shells for the coilgun that do not fit in the shelves.
It should be added that one of the bots constantly has the "cleaning" order active. The bot harvest pomegranates growing near the engine and at the same time gradually drags everything to the engineering compartment.
Regarding the fact that the bots carry spacesuits to the engineering compartment on the killer whale.
I am currently going through the campaign on the killer whale and have become convinced of the following. The bots drag everything that is not nailed to the floor to the engineering compartment. That is, everything that cannot be put in the closet. These are spacesuits of all types, empty and full boxes of shells for the magnetic gun. And I have nothing else lying around. This does not happen immediately, but gradually. At the end of the round, I find all these things in the engineering compartment under the stairs. I do not know where the bots get such a love for dragging everything. But I have to order them to ignore all the shells for the magnetic gun that do not fit in the shelves.

It should be added that one of the bots constantly has the "cleaning" order active. It collects the harvest of pomegranates growing near the engine and at the same time slowly drags everything to the engineering compartment.
This does not bother me.
Simply in response to the previously expressed doubts, I can say:
"I didn't imagine it"

**Comment by KougamiYuriko** [comment]
I've encountered one issue: when the AI is seriously injured, it ignores the pause command and actively takes medical supplies to heal itself.

**Reply by pavel4win** [reply]
IMHO that's how it's supposed to be. After all, if he dies, he won't be able to follow any orders.
But the fact that they rush at the tiger with a screwdriver, ignoring the orders "follow" and "pause", or leave the guns to fight the invaders (while there is a separate member of the team who is ordered to protect everyone), that's definitely a bug.
